compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Repl.hs

```diff
@@ -196,19 +198,33 @@ loadPackages replClient ideState = do
                 hPutStrLn stderr ("Package could not be loaded: " <> show err)
                 exitFailure
             Right _ -> pure ()
-    -- Determine module names in main DALFs.
-    md <- readMetadata (toNormalizedFilePath' ".")
+    -- Determine module names in imported DALFs.
+    let getPackage = LF.extPackagePkg . LF.dalfPackagePkg
+        unversionedPkgs = Map.fromList
+          [ (LF.packageName md, pkg)
+          | pkg@LF.Package {LF.packageMetadata = Just md} <- getPackage <$> Map.elems pkgs
+          ]
+    importLfPkgs <- forM importPkgs $ \nameOrId ->
+        let mbVersioned = getPackage <$> Map.lookup (stringToUnitId nameOrId) pkgs
+            mbUnversioned = Map.lookup (LF.PackageName . T.pack $ nameOrId) unversionedPkgs
+        in
+        case mbVersioned <|> mbUnversioned of
+            Just lfPkg -> pure lfPkg
+            Nothing -> do
+                hPutStrLn stderr $
+                    "Could not find package for import: " <> show nameOrId <> "\n"
+                    <> "Known packages: " <> intercalate ", " (unitIdString <$> Map.keys pkgs)
+                exitFailure
     pure
       [ simpleImportDecl . mkModuleName . T.unpack . LF.moduleNameString $ mod
-      | dep <- directDependencies md
-      , let pkg = LF.extPackagePkg $ LF.dalfPackagePkg $ pkgs Map.! dep
+      | pkg <- importLfPkgs
       , mod <- NM.names $ LF.packageModules pkg
       ]
 
 
-runRepl :: Options -> ReplClient.Handle -> IdeState -> IO ()
-runRepl opts replClient ideState = do
-    imports <- loadPackages replClient ideState
+runRepl :: [String] -> Options -> ReplClient.Handle -> IdeState -> IO ()
```

Comment:
If we do make the change wtr to the package/version stuff mentioned above, I would suggest to push it through here as well and parse at the edge.